packages/aws-cdk/lib/util/asset-publishing.ts

```diff
@@ -66,16 +68,32 @@ class PublishingAws implements cdk_assets.IAws {
   /**
    * Get an SDK appropriate for the given client options
    */
-  private sdk(options: cdk_assets.ClientOptions): Promise<ISDK> {
+  private async sdk(options: cdk_assets.ClientOptions): Promise<ISDK> {
+    // if this was `JSON.stringify(options)`, extra fields might
+    // accidentally be included in the cache key
+    const cacheKey = JSON.stringify({
+      region: options.region,
```

Comment:
You should use the `env` object below to index into the cache.

Reply:
I've added the "env" object in place of "region". We still need to cache based on the role information, otherwise it's possible to be incorrectly given an SDK client that assumed the file publishing role when an SDK client with asset publishing permissions is needed, or vice versa.